### ISSUE TITLE
Thread safety of MethodManager

### ISSUE THREAD

**Issue by JanPlitzkow** [sgieseking/anyrpc#27]
Hi,
I've run into a threading issue with the method manager, which was hard to find but can easily be forced to occur. First I'll give a short description how to reproduce the error, afterwards I'll explain and give my thoughts.

**Reproduction**

1.  Use the `exampleServer` and `exampleClient` from anyrpc examples. I used the single threaded JsonTcpServer, but the error probably occurs for all threaded servers.
2. In method.cpp extend the `ExecuteMethod` like this:
```
    [...]
    if (it == methods_.end())
        return false;
    std::cout << "MethodManager paused..." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(15));
    std::cout << "MethodManager going on." << std::endl;
    it->second->Execute(params,result);
    [...]
```
3. In exampleServer.cpp change the line `this_thread::sleep_for( chrono::seconds(timeout) );` to the following lines:
```
    this_thread::sleep_for( chrono::seconds(10) );
    std::cout << "Remove subtract now..." << std::endl;
    methodManager->RemoveMethod("add");
    std::cout << "Server changes finished." << std::endl;
    this_thread::sleep_for(chrono::seconds(30));
    std::cout << "Server shutting down..." << std::endl;
```
4. Start the server and immediatly afterwards the client. The client will get plenty of timeouts, but this doesn't matter. On the server you'll get at "map/set iterator not dereferencable" error (at least for VS2015).

**Explanation and Thoughts**
Inside the server's worker thread incoming rpc-commands from client are directed to the ExecuteMethod function. However this access is not protected by any mutex, so nothing prevents the main thread (or any other thread) from accessing MethodManager. By adding the sleep in ExecuteMethod, I'm just "increasing" the possibility of a thread-change at this point. 

If at this point the RemoveMethod function is called from main thread, the issue is obvious. The iterator is invalidated (and besides we do not have the corresponding function that should be executed anymore).

Note, that this issue does occur for any MethodManager function, because they all use iterators which might be invalidated. However (if I see this correctly) the invalidation does only occur for RemoveMethod. In AddMethod this issue does not occur, because a map is used and map-iterators aren't invalidated on insertion (unlike e.g. vectors). Still you could think of other problems, like adding multiple methods at the same time, listing and adding at the same time, ...

In general I would propose the usage of a mutex, which prevents from multiple access to any of the member functions, but I wanted to hear your oppinion first. The following problems might occur:

- I'm not sure if a mutex might collide with any server type. Would there be a problem, if the ExecuteMethod-function was mutexed (e.g. because it is called multiple times in parallel)? If so, one could maybe use the std::shared_mutex (C++17)...
- The call of member-functions would not be possible from within any other member-function. E.g. it would not be possible inside a rpc-command to ask for a list of all available commands. One could use a std::recursive_mutex (C++11) to allow this, but recursive mutexes are a controversial topic...

Any comments on these topics?

PS: I'm aware that the RemoveMethod function was added by one of my pull requests, but without it one would reactivate the problem of deleted references (discribed in #24).

**Comment by sgieseking** [comment]
I agree that the MethodManager class is not thread safe.  It was envisioned as something where all of the methods were created in a single thread before the server was started and then deleted when exiting so no mutex was needed while a threaded server was running.

At first I was thinking that a standard mutex could be used, but the problem is that, in the general case, the RemoveMethod function can delete the method not just remove it from the list.  So, if any other thread (or the originating thread) was executing this method, the results are undefined/dangerous.  Even with a shared_mutex, you would either need to a rule regarding self-removal, detect this situation and throw an exception, or create a deferred removal mechanism (mark for removal and check for this case after returning from the Execute function).

I might consider a solution for inclusion in the code but it would need to be elegant and low overhead.  Currently, I don’t see a good solution that meets these criteria especially when I consider it an unusual case where an application even needs to dynamically remove methods.

**Comment by sgieseking** [comment]
You might try something like the attached changes.  This uses a mutex with a counter for threads that are actively using a method.  Removing a method is deferred until after execution if there are any active threads.  I haven't really tested much so consider it a potential starting point.

[MethodManagerMutex.tar.gz](https://github.com/sgieseking/anyrpc/files/720290/MethodManagerMutex.tar.gz)

**Comment by JanPlitzkow** [comment]
Thanks for your ideas! Your solution seems to solve both of my above mentioned problems (parallel/shared access of execute and recursive call).

However the problem of #24 does arise again: Before one can delete an object, which is referenced/used by a server-method, one has to make sure this method can't be called anymore. If RemoveMethod (in some cases) only marks a method to be deleted later, this can't be assured. I would propose two versions of the RemoveMethod: One that might delay the removal (to allow self destruction) and one that assures destruction before returning (obviously recursive call not possible). The last one is possible by using conditional variable to rewake the RemoveMethod after last Execute of corresponding method is finished. One could...
1. ... have two different methods: RemoveMethod and RemoveMethodNoDelay
2. ... have an additional parameter: RemoveMethod(std::string const& name, bool allowDelay = true)

Any preferences here (I'd have a small preferece on 2)? Any other suggestions?

Besides different smaller remarks:
- I would use std::unique_lock or std::lock_guard instead of directly locking the mutex.
- The lock in destructor isn't of any use. If methodmanager is deleted, there shouldn't be any thread working with it, no matter what...
- Execute might throw, should the exeptions be handled and rethrown? After all the number of active threads should be kept updated?!

**Comment by sgieseking** [comment]
I don't see a case where the current implementation of RemoveMethod and ExecuteMethod does not work as well as it can.

If there are no active threads executing the method, then the method is immediately removed and destroyed.
If there are active threads executing the method, then it is marked as a delayed removal.  No new threads can execute the method due to the check in ExecuteMethod (it will find the method but return false).  When the last thread that is executing the method returns, the method will be removed and destroyed.  You can't do this until all of the threads have returned.  This implementation also allows the method to remove itself since it will set a delayed removal and then get removed after returning from Execute.

On the use of std::unique_lock, this seems reasonable to clean up the code and to better allow for exceptions.  This was used in ServerTP but I wasn't thinking about it when writing this code.

Agreed that the lock in the destructor shouldn't be necessary.  It would be a programming error if there are any other threads accessing the MethodManager or the methods at that point.

It's probably a good idea to catch exceptions from the Execute call and rethrow.  The delayed removal should still be called even with an exception so it may be better to put that code in a private function.

**Comment by JanPlitzkow** [comment]
> I don't see a case where the current implementation of RemoveMethod and ExecuteMethod does not work as well as it can.

Consider the following situation (as a "User" of the anyrpc classes):
- I instantiate a anyrpc-server and use different modules to provide various functionality for the clients. The different modules aren't related, except they mustn't use the same command-names (which can easily be archived by "module.command" naming scheme).
- Each module gets a pointer to the anyrpc-server, because it adds its specifiy methods. The methods are closely related to the module, they might change internal variables, data buffers, ...  
- Now I want to stop one of my modules, while clients might be using the server. Obviously the methods of this module won't be available afterwards, but I want to keep the other modules running.
- Before stopping (destructing) this module, I've to make sure its variables, data buffers, etc. are not accessed anymore. That's why RemoveMethod() is nessessary. However only knowing that the method will be removed anytime in the future is not sufficient in this case. I have to wait with my module stop / destruction until anyrpc is finished removing. Only now it's save to remove.

Sorry, for the longish answer, I hope it clarifies my intention. As far as I see, my proposal isn't that big of a difference in code. So I'll just implement the whole package including your solution and my changes and submit the code. Maybe my intention becomes clear afterwards.

Apart from the above I agree with all your statements.
